### ISSUE TITLE
alias including a newline is not highlighted anymore

### ISSUE THREAD

**Issue by itchyny** [zsh-users/zsh-syntax-highlighting#677]
With the config
```zsh
alias foobar='echo foo;
  echo bar'
```
`foobar` at the head of the prompt should be highlighted like a command but not highlighted. This used to work before the commit 9cc0060.

**Comment by phy1729** [comment]
Note that
```
alias foobar='echo foo
  echo bar'
```
will result in `foobar` being highlighted in green again.

The issue is that https://github.com/zsh-users/zsh-syntax-highlighting/blob/e2e97dde9c92bf27d8222482a08e2b77d66ac334/highlighters/main/main-highlighter.zsh#L694-L700 doesn't distinguish between `; ;` (which is likely a typo) and `;` followed by a newline (which while unnecessary doesn't indicate a likely typo).

**Comment by itchyny** [comment]
For reference, another failure case
```zsh
alias foobar='git status ||
  echo failure'
```

**Comment by danielshahaf** [comment]
Thanks for the report and the diagnosis.

Triage: This is not a recent regression, there's an easy workaround (two, actually: you could also add a `:` after the semicolon), and 0.7.0 is overdue; therefore I'm leaning towards not holding the release for this.  (We could always do a 0.7.1.)

#651 adds code to distinguish between semicolons and newlines, which we could use here.  I'll see what I can do about that.  I guess the order of business is: release 0.7.0, merge #651, [factor out the detection of `$'\n'`](https://github.com/zsh-users/zsh-syntax-highlighting/pull/651#discussion_r365502788) if possible, fix this issue.  Does this sound good?

@itchyny Naturally, if you'd like to look into a fix, you'd be welcome to.

**Comment by danielshahaf** [comment]
Documented in changelog as a known issue.

**Comment by danielshahaf** [comment]
Isn't this the same issue as #616?

**Comment by danielshahaf** [comment]
> Isn't this the same issue as #616?

It's related but not the same.  Compare these two tests:

https://github.com/zsh-users/zsh-syntax-highlighting/blob/54e1828d5c515730f3d9fb026a8bf9e8eb2cc970/highlighters/main/test-data/alias-comment1.zsh#L30-L37
https://github.com/zsh-users/zsh-syntax-highlighting/blob/54e1828d5c515730f3d9fb026a8bf9e8eb2cc970/highlighters/main/test-data/comment-followed.zsh#L31-L38

With #651 having been merged, the second test sees `arg=$'\n'` but the first test does not, because [the alias parsing codepath runs ${(z)} separately](https://github.com/zsh-users/zsh-syntax-highlighting/blob/54e1828d5c515730f3d9fb026a8bf9e8eb2cc970/highlighters/main/main-highlighter.zsh#L603-L609).

To complicate things, there's another test that calls itself #616:

https://github.com/zsh-users/zsh-syntax-highlighting/blob/54e1828d5c515730f3d9fb026a8bf9e8eb2cc970/highlighters/main/test-data/empty-command-newline.zsh#L31-L39

So:

- [ ] Update alias-comment1.zsh to refer to this issue.  Keep empty-command-newline.zsh referring to #616.
- [ ] Ask zsh upstream to add a ${(Z)} mode that preserves newlines as newlines, so we don't have to re-implement the a4525a0826a3e1b0fc3def19a0a4a527c7045f3d dance everywhere we use ${(z)}, such as in the aliases code.
- [ ] Fix #616 and #501 using #561.
- [ ] To fix this issue, we'll need to re-apply a4525a0826a3e1b0fc3def19a0a4a527c7045f3d to [the ${(z)} use in the alias-parsing code](https://github.com/zsh-users/zsh-syntax-highlighting/blob/54e1828d5c515730f3d9fb026a8bf9e8eb2cc970/highlighters/main/main-highlighter.zsh#L603-L609), or to implement the ${(Z)} subflag and wait for a zsh release that includes it.  (As of this writing, the latest zsh release is 5.8.)

**Comment by danielshahaf** [comment]
A third workaround is to convert the alias to a function.  In the given example that's probably a good idea anyway, since functions can't misfire when used in SHORT_LOOPS loop bodies.  (`alias x='date; pwd'; repeat 2 x`)
